src/game/msg_builder.cc

```diff
@@ -0,0 +1,82 @@
+#include "msg_builder.h"
+
+namespace SailGame { namespace Game {
+
+ProviderMsgPtr MsgBuilder::CreateRegisterArgs(int seqId, const std::string &id,
+    const std::string gameName, int maxUsers, int minUsers)
+{
+    ProviderMsg msg;
+    msg.set_sequenceid(seqId);
+
+    auto registerArgs = msg.mutable_registerargs();
+    registerArgs->set_id(id);
+    registerArgs->set_gamename(gameName);
+
+    auto gameSettings = registerArgs->mutable_gamesetting();
+    gameSettings->set_maxusers(maxUsers);
+    gameSettings->set_minusers(minUsers);
+
+    return std::make_shared<ProviderMsg>(msg);
```

Comment:
I guess there is only one thread?
how about define 'ProviderMsg mMsg'  in MsgBuilder.
And just use that member, and return ProviderMsg* ?

Reply:
I think std::make_shared<ProviderMsg>(msg); would trigger copy-constructor.

Reply:
actually, if you are using multiple threads,  just let each thread holds its own MsgBuilder.

Reply:
yep there is only one thread and it's a great idea to define msg as member, however the return value of `Transition` is a vector of `ProviderMsgPtr`. That is to say, in a transition there might be multiple `ProviderMsg`s needed to send. so the optimization doesn't apply to our scenario.

Reply:
as for the copy constructor triggered by `std::make_shared` return value, that's true. already fixed.